Extension/src/LanguageServer/utils.ts

```diff
@@ -1,101 +1,108 @@
+    return cpptoolsTextEdits.map(textEdit => new vscode.TextEdit(makeVscodeRange(textEdit.range), textEdit.newText));
+}
+
+export function rangeEquals(range1: vscode.Range | Range, range2: vscode.Range | Range): boolean {
+    return range1.start.line === range2.start.line && range1.start.character === range2.start.character &&
+    range1.end.line === range2.end.line && range1.end.character === range2.end.character;
+}
+
+// Check this before attempting to switch a document from C to C++.
+export function shouldChangeFromCToCpp(document: vscode.TextDocument): boolean {
+    if (document.fileName.endsWith(".C") || document.fileName.endsWith(".H")) {
+        const cppSettings: CppSettings = new CppSettings();
+        if (cppSettings.autoAddFileAssociations) {
+            return !docsChangedFromCppToC.has(document.fileName);
+        }
+        // We could potentially add a new setting to enable switching to cpp even when files.associations isn't changed.
+    }
+    return false;
+}
+
+// Call this before changing from C++ to C.
+export function handleChangedFromCppToC(document: vscode.TextDocument): void {
+    if (shouldChangeFromCToCpp(document)) {
+        docsChangedFromCppToC.add(document.fileName);
+    }
+}
+
+export function showInstallCompilerWalkthrough(): void {
+    // Because we need to conditionally enable/disable steps to alter their contents,
+    // we need to determine which step is actually visible. If the steps change, this
+    // logic will need to change to reflect them.
+    enum Step {
+        Activation = 'awaiting.activation',
+        NoCompilers = 'no.compilers.found',
+        Verify = 'verify.compiler'
+    }
+
+    const step = (() => {
+        if (!SessionState.scanForCompilersDone.get()) {
+            return Step.Activation;
+        } else if (!SessionState.trustedCompilerFound.get()) {
+            return Step.NoCompilers;
+        } else {
+            return Step.Verify;
+        }
+    })();
+
+    const platform = (() => {
+        switch (os.platform()) {
+            case 'win32': return 'windows';
+            case 'darwin': return 'mac';
+            default: return 'linux';
+        }
+    })();
+
+    const version = (platform === 'windows') ? SessionState.windowsVersion.get() : '';
+
+    const index = `ms-vscode.cpptools#${step}.${platform}${version}`;
+
+    void vscode.commands.executeCommand(
+        'workbench.action.openWalkthrough',
+        { category: 'ms-vscode.cpptools#cppWelcome', step: index },
+        false)
+        // Run it twice for now because of VS Code bug #187958
+        .then(() => vscode.commands.executeCommand(
+            "workbench.action.openWalkthrough",
+            { category: 'ms-vscode.cpptools#cppWelcome', step: index },
+            false)
+        );
+    return;
+}
+
+const docsChangedFromCppToC: Set<string> = new Set<string>();
+
+export async function withCancellation<T>(promise: Promise<T>, token: vscode.CancellationToken): Promise<T> {

```

Comment:
Does the call to `token.onCancellationRequest` permanently modify the token passed in, to add the callback?  In general, a helper like this would be useful in some of our providers, but it seems like a bug that the callback registration is not only within the scope of this call.  We would also need to be able to call `onCancellationRequested` after calling this, to register other work we may need to do on cancellation.  I'm not sure of the consequences of multiple calls to `onCancellationRequested`.

Reply:
@Colengms if I understand correctly, it is more about stylistic approach or future refactoring, but in principle no changes are required now in this PR. Let me know.